### PR TITLE
🌱 hack: add capv-janitor for automated ci cleanup

### PR DESCRIPTION
**Pull request by chrischdi**
**What this PR does / why we need it**:

Adds `hack/tools/janitor`, a small tool which can be run as cronjob in CI to cleanup leftover objects over time.

As of now cleanup was done manually but since running jobs in parallel, we may have more leftover objects so its more often required to cleanup things.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

**Comment by chrischdi** [comment]
/retest

(flake)

**Comment by sbueringer** [comment]
I'll take a look next week

File: hack/tools/janitor/vSphere.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+	vimClient, err := vim25.NewClient(ctx, soapClient)
+	if err != nil {
+		return nil, err
+	}
+
+	govmomiClient := &govmomi.Client{
+		Client:         vimClient,
+		SessionManager: session.NewManager(vimClient),
+	}
+	// To keep the session from timing out until the test suite finishes
```
Comment (sbueringer) [inline hack/tools/janitor/vSphere.go:87]: Is this a realistic problem? I thought the timeout duration is something like 30m?

(also we'll be continuously doing requests)

Comment (sbueringer) [reply hack/tools/janitor/vSphere.go:87]: Just want to avoid dealing with this stuff if its' not necessary

File: hack/tools/janitor/janitor.go [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+	"github.com/vmware/govmomi/vim25/mo"
+	"github.com/vmware/govmomi/vim25/types"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/klog/v2"
+	ipamv1 "sigs.k8s.io/cluster-api/exp/ipam/api/v1alpha1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func newJanitor(vSphereClients *vSphereClients, ipamClient client.Client, delta time.Duration, ipamNamespace string, dryRun bool) *janitor {
```
Comment (sbueringer) [inline hack/tools/janitor/janitor.go:37]: Let's rename the "delta" arg (maybe to minAge?)

File: hack/tools/janitor/janitor.go [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+			// Skipping actual PowerOff on dryRun.
+			continue
+		}
+		task, err := vm.object.PowerOff(ctx)
+		if err != nil {
+			return err
+		}
+		poweroffTasks = append(poweroffTasks, task)
+	}
+	// Wait for all PowerOff tasks to be finished. We intentionally ignore errors here
```
Comment (sbueringer) [inline hack/tools/janitor/janitor.go:129]: So this is overall a best effort powerOff so they give back their IPs? (let's add a godoc with some context around why we are trying poweroff)

File: hack/tools/janitor/janitor.go [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+		}
+		task, err := vm.object.Destroy(ctx)
+		if err != nil {
+			return err
+		}
+		destroyTasks = append(destroyTasks, task)
+	}
+	// Wait for all destroy tasks to succeed.
+	if err := waitForTasksFinished(ctx, destroyTasks, false); err != nil {
+		return errors.Wrap(err, "failed to wait for vm poweroff task to finish")
```
Comment (sbueringer) [inline hack/tools/janitor/janitor.go:151]: Strange copy&paste error :D (it's not the same but seems to be from poweroff)

File: hack/tools/janitor/janitor.go [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+	}
+
+	return nil
+}
+
+func waitForTasksFinished(ctx context.Context, tasks []*object.Task, ignoreErrors bool) error {
```
Comment (sbueringer) [inline hack/tools/janitor/janitor.go:157]: Would it make sense to print all tasks somewhere? So if we already return after the first failed we have a list of taskIDs we can use to look into the problem?

File: hack/tools/janitor/janitor.go [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+		if err := s.ipamClient.Delete(ctx, &ipAddressClaim); err != nil {
+			errList = append(errList, err)
+		}
+	}
+
+	return kerrors.NewAggregate(errList)
+}
+
+func (s *janitor) vSphereClusterModules(ctx context.Context) error {
+	log := ctrl.LoggerFrom(ctx).WithName("vSphere cluster modules")
```
Comment (sbueringer) [inline hack/tools/janitor/janitor.go:206]: Is there any way we can consider maxCreationDate here?

I'm wondering if this code here could theoretically delete cluster modules an e2e test just created

**Comment by sbueringer** [comment]
Main comment is this one: https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/2681#discussion_r1478065626

(potentially a lot follows from that)

File: hack/tools/janitor/config/cronjob/cronjob.yaml [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+          - args:
+            - --dry-run=false
+            - --min-age=12h
+            # In CAPV's CI IPAddressClaims are created in the default namespace.
+            - --ipam-namespace=default
+            # The directories used in CI to cleanup.
+            - --folder=/SDDC-Datacenter/host/Cluster-1/Resources/Compute-ResourcePool/cluster-api-provider-vsphere
+            - --folder=/SDDC-Datacenter/host/Cluster-1/Resources/Compute-ResourcePool/cloud-provider-vsphere
+            - --folder=/SDDC-Datacenter/host/Cluster-1/Resources/Compute-ResourcePool/image-builder
+            - --folder=/dc0/vm/folder0/cluster-api-provider-vsphere
```
[the remark this answers is not in the document]
Comment (chrischdi) [reply hack/tools/janitor/config/cronjob/cronjob.yaml:26]: ack, forgot to remove this one :-) thx!

File: hack/tools/janitor/janitor.go (hunk 6 above)
Comment (sbueringer) [reply hack/tools/janitor/janitor.go:206]: Okay then let's just see if we run into weird flakes

File: hack/tools/janitor/vSphere.go (hunk 1 above)
Comment (chrischdi) [reply hack/tools/janitor/vSphere.go:87]: Was copy & paste but ok to remove it 👍

File: hack/tools/janitor/janitor.go (hunk 3 above)
Comment (chrischdi) [reply hack/tools/janitor/janitor.go:130]: nope, thats a hard poweroff which is required before deleting a VM.

I think we're good with the dhcp lease duration :-)

Comment (sbueringer) [reply hack/tools/janitor/janitor.go:134]: I meant best effort because we only log errors in l.138 :)

File: hack/tools/janitor/janitor.go [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+	maxCreationDate time.Time
+	vSphereClients  *vSphereClients
+}
+
+type virtualMachine struct {
+	managedObject mo.VirtualMachine
+	object        *object.VirtualMachine
+}
+
+// deleteVSphereVMs deletes all deleteVSphereVMs in a given folder in vSphere if their creation
```
Comment (sbueringer) [inline hack/tools/janitor/janitor.go:60]:
```suggestion
// deleteVSphereVMs deletes all VSphereVMs in a given folder in vSphere if their creation
```

File: hack/tools/janitor/janitor.go (hunk 3 above)
Comment (sbueringer) [reply hack/tools/janitor/janitor.go:134]: So I think the open question is. Should we fail here or is it fine if we just log if an error occurs

**Comment by sbueringer** [comment]
/lgtm
/appprove

/hold

Not sure if you want further reviews.

**Pull request merged**